File: dart/examples/marketplace_eligibility/lib/src/projecting_rule.dart
import 'package:verdict_rules/verdict_rules.dart';

/// Adapts a `Rule<TInner>` to run inside a composite built on `TOuter`.
///
/// Not part of verdict_rules itself -- a consumer-defined adapter, exactly
/// as free to exist as a new rule shape is, with no changes needed on
/// verdict_rules' side to support it. See
/// docs/extending/reusing-a-rule-across-contexts/dart.md for the standalone
/// version of this same pattern.
class ProjectingRule<TOuter, TInner> implements Rule<TOuter> {
  @override
  final String name;

  @override
  final String? group;

  final Rule<TInner> _inner;
  final TInner Function(TOuter) _project;

  ProjectingRule(Rule<TInner> inner, TInner Function(TOuter) project)
      : name = inner.name,
        group = inner.group,
        _inner = inner,
        _project = project;

  @override
  Future<RuleResult> evaluate(TOuter context) =>
      _inner.evaluate(_project(context));
}

File: csharp/examples/MarketplaceEligibility/ProjectingRule.cs
using VerdictRules;

namespace MarketplaceEligibility;

/// <summary>
/// Adapts an <see cref="IRule{TInner}"/> to run inside a composite built on
/// <typeparamref name="TOuter"/>.
/// </summary>
/// <remarks>
/// Not part of verdict-rules itself. See
/// docs/extending/reusing-a-rule-across-contexts/csharp.md.
/// </remarks>
public sealed class ProjectingRule<TOuter, TInner>(
    IRule<TInner> inner,
    Func<TOuter, TInner> project) : IRule<TOuter>
{
    public string Name { get; } = inner.Name;
    public string? Group { get; } = inner.Group;

    public Task<RuleResult> EvaluateAsync(TOuter context, CancellationToken cancellationToken = default) =>
        inner.EvaluateAsync(project(context), cancellationToken);
}
